Canvases - Part Deux (Loading/Saving + Frontend API)

dsaedtler opened the pull request:
### Description

Adds saving and loading for frontend-owned canvases as well as a frontend API to create/remove/enumerate them.

### Motivation and Context

In order to persist additional canvases and their scenes/sources in a way that is reasonably straightforward for API users we want the frontend to own/manage additional canvases that it may interface with.

In the future I'd also like to make transitions and the main canvas part of this abstraction.

### How Has This Been Tested?

This was tested with this fork of the vertical canvas plugin that makes use of the new APIs: https://github.com/dsaedtler/obs-vertical-canvas

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

RytoEX commented on frontend/api/obs-frontend-api.h:
```diff
@@ -80,6 +83,18 @@ static inline void obs_frontend_source_list_free(struct obs_frontend_source_list
+static inline void obs_frontend_canvas_list_free(struct obs_frontend_canvas_list *canvas_list)
+{
+	size_t num = canvas_list->canvases.num;
+	for (size_t i = 0; i < num; i++)
+		obs_canvas_release(canvas_list->canvases.array[i]);
+	da_free(canvas_list->canvases);
```
Add a blank line before the `da_free` to separate it from the `if`, or enclose the `if` body in curly braces.

RytoEX commented on frontend/widgets/OBSBasic.hpp:
```diff
@@ -1109,6 +1113,27 @@ private slots:
+
+	static void CanvasRemoved(void *data, calldata_t *params);
+
+public slots:
+	bool RemoveCanvas(OBSCanvas canvas);
+
+public:
+	const std::vector<OBSFrontendCanvas> &GetCanvases() const noexcept { return canvases; }
+
+	OBSCanvas AddCanvas(const std::string &name, obs_video_info *ovi = nullptr, int flags = 0);
+
```
Could these be ordered Add, Remove, Get like in the frontend API files?

dsaedtler replied:
I reordered everything to get Get, Add, Remove everywhere.

PatTheMav reviewed (commented):
I find the whole mix and match of C types by `libobs`, "native" C++ types used by the front-end, and "RAII"-wrapper types unfortunate, particularly the back and forth between `OBSFrontendCanvas`, `OBSCanvas`, and `obs_canvas_t`.

I would prefer if Canvas management had been properly encapsulated into its own class (which would also be the owner of the data and pointers) which could potentially make using the RAII wrappers as well as a wrapper around the wrapper for the vector superfluous.

Free standing functions could then serve as factory methods that yield either a vector of Canvas objects or an `obs_data_array_t` of canvas data. And `OBSBasic` would just call something like `std::vector<OBS::Canvas> OBS::Canvas::LoadCanvases(obs_data_array_t *canvases)` which would elide a copy thanks to RVO.

And instead of trying to get the pointer wrapped by the RAII wrapper and calling `obs_canvas_remove` on each of them, one would be call `OBS::Canvas::Erase()` or similar for each item in the vector.

---

Apart from my architectural concerns, the code as-is looks correct. But if we introduce a whole new thing like canvases into the program, I'd like us to do it right and use appropriate encapsulation and abstraction.

dsaedtler commented on frontend/OBSStudioAPI.cpp:
```diff
@@ -682,6 +682,26 @@ void OBSStudioAPI::obs_frontend_add_undo_redo_action(const char *name, const und
+	for (auto &fc : main->canvases) {
+		obs_canvas_t *canvas = obs_canvas_get_ref(fc.canvas);
+		if (canvas)
+			da_push_back(canvas_list->canvases, &canvas);
```
Strong references.

dsaedtler commented on frontend/widgets/OBSBasic_Canvases.cpp:
```diff
@@ -0,0 +1,96 @@
+		return false;
+
+	auto canvas_it = std::find_if(std::begin(canvases), std::end(canvases),
+				      [&](const OBSFrontendCanvas &fc) { return fc.canvas.Get() == canvas.Get(); });
```
I don't think it would necessarily be safer, but it would be relatively easy to do at least (just create a `const string_view` with the UUID and then compare it in the lambda).

PatTheMav commented:
> > I find the whole mix and match of C types by `libobs`, "native" C++ types used by the front-end, and "RAII"-wrapper types unfortunate, particularly the back and forth between `OBSFrontendCanvas`, `OBSCanvas`, and `obs_canvas_t`.
> > I would prefer if Canvas management had been properly encapsulated into its own class (which would also be the owner of the data and pointers) which could potentially make using the RAII wrappers as well as a wrapper around the wrapper for the vector superfluous.
> > Free standing functions could then serve as factory methods that yield either a vector of Canvas objects or an `obs_data_array_t` of canvas data. And `OBSBasic` would just call something like `std::vector<OBS::Canvas> OBS::Canvas::LoadCanvases(obs_data_array_t *canvases)` which would elide a copy thanks to RVO.
> 
> That sounds fine, the `OBSFrontendCanvas` was originally intended to also carry transitions (to save/load them for non-primary canvases, and eventually the main one as well), but I didn't want to do that for this first iteration. But it could certainly be extended into a full-on wrapper that provides a more OOP API.
> 
> Where do you think such class should live in the current code structure? `frontend/utility/OBSCanvas.{hpp,cpp}`?

Yeah `utility` would be the "fallback" if we cannot come up with a better category for these things. In the future we'd need some directory for bespoke classes for transitions, scenes, scene collections, which aren't "utility" per-se, but core building blocks of the compositor.

dsaedtler commented:
I've made some changes based on the above conversation (the diff may be a bit messy due to rebasing, sorry):

- Added an `OBS::Canvas` class, that owns `obs_canvas_t` objects
    + Supports move semantics for `std::vector`, which automatically calls `obs_canvas_remove()` on the owned canvas when erased
- Removed `OBSFrontendCanvas`
- Created two free standing functions for loading/saving a vector of `OBS::Canvas` objects from `obs_data_array_t`

I'm not sure where the freestanding functions should be, so right now they're just part of the `OBSCanvas.cpp` file as `OBS::LoadCanvases` and `OBS::SaveCanvases`.

PatTheMav commented on frontend/widgets/OBSBasic_Canvases.cpp:
```diff
@@ -0,0 +1,50 @@
+	QMetaObject::invokeMethod(static_cast<OBSBasic *>(data), "RemoveCanvas", Q_ARG(OBSCanvas, OBSCanvas(canvas)));
+}
+
+OBSCanvas OBSBasic::AddCanvas(const std::string &name, obs_video_info *ovi, int flags)
```
Shouldn't this be `OBS::Canvas`?

PatTheMav replied:
@dsaedtler This should be the last open thread on this PR.

dsaedtler replied:
Addressed in the latest push.

PatTheMav commented on frontend/widgets/OBSBasic_Canvases.cpp:
```diff
@@ -0,0 +1,50 @@
+	return canvas;
+}
+
+bool OBSBasic::RemoveCanvas(obs_canvas_t *canvas)
```
With `canvases` being a collection of `OBS::Canvas` instances, shouldn't this instead get a name or UUID and then remove inside the collection based on that?

You can't find the corresponding `OBS::Canvas` instance based on the underlying C object pointer (it should not even be part of the frontend code, see above) and you can't send an object instance around to delete based on identity (which would create all kinds of lifetime problems).

dsaedtler replied:
This currently primarily exists to get called by the `canvas_remove` signal from `libobs`, which like `source_remove` passes a pointer to the object. But it could be made private and a public API that takes a UUID instead could be added.

PatTheMav replied:
My bad, I overlooked the `operator obs_canvas_t *() const` operator on the Canvas class, so the `std::find` call will work as expected.

PatTheMav reviewed (commented):
Apart from my comments and to ensure this doesn't get lost: https://github.com/obsproject/obs-studio/pull/11824 needs to be merged first and this PR then based on top of it, as it introduces changes to the scene collections implementation on `OBSBasic`.

dsaedtler commented on frontend/widgets/OBSBasic_Canvases.cpp:
```diff
@@ -0,0 +1,50 @@
+	if (name.empty())
+		return nullptr;
+
+	OBSCanvas canvas = obs_canvas_create(name.c_str(), ovi, flags);
```
These functions primarily exist as adapters to the frontend API, which ultimately needs to return an `obs_canvas_t *` rather than any higher level type since it's a C API. Should the frontend API code here handle the transition from `OBS::Canvas` to `obs_canvas_t*` while the frontend itself remains oblivious to the underlying `libobs` object?

PatTheMav replied:
Just so I understand this right: Canvases are implemented as first-class `libobs` types, so why does a plugin have to do a round trip over the frontend to interact with them?

If I'm not mistaken this design requires the flow of
```
plugin -> frontend API -> frontend -> libobs
```
even though it could just be
```
plugin -> libobs
```
Why is the frontend involved if canvasses (their existence as well as their management) is implemented in `libobs` in the first place? The frontend needs to have a way to interact with the canvases (which the C++ wrapper class is for), but to use the frontend(!) API to allow plugins to interact with seems like a big detour to me, or am I'm understanding this wrong?

dsaedtler replied:
To clarify why this exists in this manner, of course a plugin can create a canvas directly and use it however it wants. But based on our discussions a few months ago the frontend should manage any canvases that can be used in its outputs (e.g. multitrack) and should also handle persistence. So the frontend API exists to create a frontend-managed canvas that gets saved/loaded to/from the scene collection automatically, rather than the plugin having to worry about the canvas and its scenes+sources itself.

And once OBS is able to manage multiple canvases natively (something I know @Warchamp7 wants to do and we might help with if desired) the API allows plugins to interact with the OBS-created and owned canvases as well.

PatTheMav replied:
So the frontend API is supposed to give plugins access to the canvases created and owned _by_ the frontend and it's not just "another way" to access the same list of canvases that might be registered with `libobs`?

dsaedtler replied:
Correct. The frontend only knows/cares about (and persists) canvases created by itself and/or by plugins via the frontend API. All other canvases exists solely in `libobs` space and have to be managed by the API user themselves.

Effectively the list of canvases in the frontend should be considered to be part of the scene collection. I know you've done some refactoring there recently that I have not yet looked at in detail, so perhaps there is a better place we can put this now or in the future.

PatTheMav replied:
That should be doable at a later step - the first step was to make scene collections a class type whose instance references can be shared and relevant data can be migrated into those instances later.

So the collection as well as the relevant APIs would then be moved into the scene collection class, but that's for later.

dsaedtler commented on frontend/utility/OBSCanvas.cpp:
```diff
@@ -0,0 +1,103 @@
+	return std::nullopt;
+}
+
+std::vector<Canvas> LoadCanvases(obs_data_array_t *canvases)
```
This is now a static member as well.
